### ISSUE TITLE
Retry bugdet implementation

### ISSUE THREAD

**Issue by jansyk13** [resilience4j/resilience4j#172]
It would be cool to implement retry budget policy for retries, current retry policy can cause retry hell. We would like to use it for retrying requests to our downstream services. I know that combination of retry and circuit breaker could have similar effects. Example of the policy is described here https://finagle.github.io/blog/2016/02/08/retry-budgets/ and in their guide there is more information.

I could prepare some POC if you would be interested.

**Comment by RobWin** [comment]
Hi, interesting idea. Indeed, a retry bugdet solves a different issue than a circuit breaker. Feel free to prepare a PR.

**Comment by storozhukBM** [comment]
Great idea @jansyk13, looking forward for it.

**Comment by gibffe** [comment]
Any eta on rx operator for this ? :) Love the idea.

**Comment by jansyk13** [comment]
@gibffe still in a process of figuring out, how it should work...
My proposition:
* use RingBitSet (implemented in r4j)
* 3 params: (ring size, % of allowed retries, max # retries per execution)
* onFinishedExecution - mark false in set(both success and failure), maybe mark false could be only onSuccess
* onFailureExecution - check if retry is allowed and if yes proceed and mark true in set(in case of another failure repeat until max # retries)
* check if retry is allowed = compare the number of retries in set with total executions in set
* if retry is denied then call recovery function

Cons:
* no "time"(scheduled) function that would get rid of retries
* in case of small ring(set) the retries can be flushed out quickly(it would heavily relay on correct configuration in deployment)

What do you think? cc @storozhukBM @RobWin

**Comment by storozhukBM** [comment]
I'd user RateLimiter for it.

**Comment by storozhukBM** [comment]
Where you can retry but not faster then N times per sec or per minute...

**Comment by gibffe** [comment]
I'd rather configure this in terms of % not tps (as in finagle example).
Coupling retry budget to tps is not ideal imo.

On Nov 5, 2017 7:43 AM, "Bogdan Storozhuk" <notifications@github.com> wrote:

Where you can retry but not faster then N times per sec or per minute...

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/resilience4j/resilience4j/issues/172#issuecomment-341982468>,
or mute the thread

**Comment by storozhukBM** [comment]
@jansyk13 do you have any updates or new thoughts on it?

**Comment by jansyk13** [comment]
I have prepared PR, https://github.com/resilience4j/resilience4j/pull/192

Could you please review it? @storozhukBM @RobWin @gibffe

**Comment by RobWin** [comment]
Yes thanks.
As soon as I have some time.

**Comment by RobWin** [comment]
Should we start a second try to implement a RetryBudget?

**Comment by jansyk13** [comment]
@RobWin Sure, but I think there should be consensus on the design and concepts first.

As far as I know my previous company uses the retry budget based on closed PR. It was turned into internal library and is used by API gateway. The design based on RingBitSet has limitations, but I believe that it works well for them.

**Comment by IgorPerikov** [comment]
@RobWin Hi, I'm in. Was about to create a new issue, but thanks to _similar issues functionality_ 😄

I was planning to describe possible solution as following:

From my point of view, it's important for capacity planning to limit clearly the maximum increase of load on upstreams. Idea is to keep track (via rolling window) total number of calls (whether success or not) and amount of first-try-attempts (=basically number of times retry section is being entered). Then define error budget as division of first by second. Therefore error budget will mean the maximum allowed increase in load. Logically we need to add only 1 new parameter - error budget itself and according to the definition default value would be `maxAttempts * 100%`, but I understand that it wouldn't be the best technical solution, so it's probably better to come up with 2 different strategies, or resolve them inside Builder/Factory depending on whether errorBudget was passed/set or not

I haven't looked thoroughly at previous implementation attempt and not completely familiar with resilience4j codebase, but I would like to elaborate and work on it, if you don't mind.

**Comment by RobWin** [comment]
Sure, you can work on it. I might be helpful to discuss it in our slack channel.
I will add my ideas tomorrow. I think it's simpler to create a new a `AdaptiveRetry` decorator instead of modifying the existing `Retry`. We are currently creating an `AdaptiveBulkhead` which uses an AIMD algorithm to adjust the concurrency limit based on the outcome and latency of the calls from the last N seconds. That means the concurrency limit is automatically reduced or increased to find the best concurrency limit.
I could imagine that we create an `AdaptiveRetry` which also adjust the maxAttempts variable based on the call results and latency of the last N calls or last N seconds. Might be more adaptive than a fixed rps limit or token bucket size.

**Comment by IgorPerikov** [comment]
> I might be helpful to discuss it in our slack channel.

I don't see it on repository page, should I be invited specifically?

**Comment by RobWin** [comment]
Yes, if you really would like to work on it, I can invite you, if you provide me an email address .

**Comment by IgorPerikov** [comment]
Yes, I am
perikov.igor@gmail.com

**Comment by RahulKushwaha** [comment]
@RobWin Are you guys still working on the idea? I would certainly like to contribute. 
If not I would like to start on it. (maybe revive PR, [#192](https://github.com/resilience4j/resilience4j/pull/192))

**Comment by RobWin** [comment]
No, you can pick it if you like.

**Comment by RahulKushwaha** [comment]
@RobWin That will be awesome!! Two questions: 
1. Should I assign the issue to myself?
2. I was going through the PR, comments, etc. It seems like you guys were also heading towards some solutions/discussions. Do we have any PR, document where I can read them? Otherwise, I can start from scratch.

**Comment by veravratkevich** [comment]
Hi! Could you please tell me if there are any updates on this issue?

**Comment by MPSforca** [comment]
Hello! Are there any updates regarding this Feature? I do see a lot of value on the Retry Budget as I believe it may solve a problem many of us currently have with Retries 🙂 

Could we "revive" the PR https://github.com/resilience4j/resilience4j/pull/192?

Also, I'd like to rethink about the possibility of using the RingBitSet. The implementation with this algorithm seems to be too dependent on choosing the "right" buffer size: a smaller buffer could make the retry budget too sensitive while a bigger buffer may take too long for the Retry Budget to start blocking the retries (say that we have a buffer of size 100, but during the last 10 requests the server started to have an instability. If over the last 90 requests we didn't have a retry, it just started now, the Retry Budget won't block them quickly)

I would suggest to implement the Retry Budget by keeping track of the retry ratio (`retries/requests`) over a time window. Having the implementation considering the time seems more flexible for me. 
For example, for a new Retry, we set the time window for 10 seconds for example. For a new request / retry, we store the current time (if none is defined). If it's a request, we add up to the `requests_count`. Otherwise, if it's a retry, we add up to the `retry_count`. 
For each new request or retry, we check the current time. If it has passed more than the 10 seconds of time window, we reset the counters. Otherwise, if it's a retry, we check the current ratio and if it's over the proposed "max" ratio, we block it.

What do you think?

**Comment by RobWin** [comment]
Hi @MPSforca 
We are always open for contributions.
You can revive the PR and work on it and test and show your ideas on working code.
